src/Compilers/CSharp/Portable/Symbols/Source/SourceOrdinaryMethodSymbolBase.cs

```diff
@@ -0,0 +1,671 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Globalization;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.CSharp.Emit;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols
+{
+    internal abstract class SourceOrdinaryMethodSymbolBase : SourceMemberMethodSymbol
+    {
+        private readonly ImmutableArray<TypeParameterSymbol> _typeParameters;
+        private readonly string _name;
+
+        private ImmutableArray<MethodSymbol> _lazyExplicitInterfaceImplementations;
+        private ImmutableArray<CustomModifier> _lazyRefCustomModifiers;
+        private ImmutableArray<ParameterSymbol> _lazyParameters;
+        private TypeWithAnnotations _lazyReturnType;
+
+        protected SourceOrdinaryMethodSymbolBase(
+            NamedTypeSymbol containingType,
+            string name,
+            Location location,
+            SyntaxReference syntaxReference,
+            MethodKind methodKind,
+            bool isIterator,
+            bool isExtensionMethod,
+            bool isPartial,
+            bool hasBody,
```

Comment:
Do we want to deal with this in the derived type? It would mean the derived types have to do more work, but it kind of moves the work to where it belongs -- alongside the syntax.

Not pressing, I don't think it really matters, but if we wanted a stronger separation between syntax and semantics (which I think is good and would make stuff like this easier in the future), that might be something to consider.

Reply:
> Do we want to deal with this in the derived type? It would mean the derived types have to do more work, but it kind of moves the work to where it belongs -- alongside the syntax.

As we discussed offline. At the moment, we can take advantage of code sharing for reporting diagnostics that is specific for methods that have a body.
For methods, if they have a declaration syntax, the presence of a body is a syntax check. The synthesized ones know if there is a body based on some specific situation and we want to perform the checks for them too.


---
In reply to: [445817294](https://github.com/dotnet/roslyn/pull/45424#discussion_r445817294) [](ancestors = 445817294)